lib/srv/db/server.go

```diff
@@ -920,24 +905,84 @@ func (s *Server) Shutdown(ctx context.Context) error {
 }
 
 func (s *Server) close(ctx context.Context) error {
-	var errors []error
-	// Stop proxying all databases.
-	for _, database := range s.getProxiedDatabases() {
-		if services.ShouldDeleteServerHeartbeatsOnShutdown(ctx) {
-			errors = append(errors, trace.Wrap(s.stopProxyingAndDeleteDatabase(ctx, database)))
-		} else {
-			errors = append(errors, trace.Wrap(s.stopProxyingDatabase(ctx, database)))
+	shouldDeleteDBs := services.ShouldDeleteServerHeartbeatsOnShutdown(ctx)
+	sender, ok := s.cfg.InventoryHandle.GetSender()
+	if ok {
+		// Manual deletion per database is only required if the auth server
+		// doesn't support actively cleaning up database resources when the
+		// inventory control stream is terminated during shutdown.
+		if capabilities := sender.Hello().Capabilities; capabilities != nil {
+			shouldDeleteDBs = shouldDeleteDBs && !capabilities.DatabaseCleanup
+		}
+	}
+	g, gctx := errgroup.WithContext(ctx)
+	g.SetLimit(100)
+
+	// Hold the READ lock while iterating the databases here to prevent
+	// deadlocking in flight heartbeats. The heartbeat announce acquires
+	// the lock to build the db resource to send. If the WRITE lock is
+	// held during the shutdown procedure below, any in flight heartbeats
+	// will block acquiring the mutex until shutdown completes, at which
+	// point the heartbeat will be emitted and the removal of the db
+	// server below would be undone.
+	s.mu.RLock()
+	for name := range s.proxiedDatabases {
+		name := name
+		heartbeat := s.heartbeats[name]
+
+		if dynamic, ok := s.dynamicLabels[name]; ok {
+			dynamic.Close()
+		}
+
+		if heartbeat != nil {
+			log := s.log.With("db", name)
+			log.DebugContext(ctx, "Stopping db")
+			if err := heartbeat.Close(); err != nil {
+				log.WarnContext(ctx, "Failed to stop db.", "error", err)
+			} else {
+				log.DebugContext(ctx, "Stopped db")
+			}
+
+			// Stop database object importer.
+			if err := s.cfg.DatabaseObjects.StopImporter(name); err != nil {
+				s.log.WarnContext(ctx, "Failed to stop database object importer.", "db", name, "error", err)
+			}
```

Comment:
this should be outside of `if heartbeat != nil`? Alternatively it can implement a Close or StopAll function.